DATA ACQUISITION

Data can stored locally or in data bases or in the cloud.
Data can be collected from websites like bbc, twitter etc. They come in various formats.mata.

Reading Datacan be by using Pandas or Open

In [143]:
import numpy as np
import pandas as pd

reading an excel file

excelFile = pd.read_excel("directory to file", sheet_name = "Sheet2")

ignore sheet name if not needed

csv file

csvFile = pd.read_csv("dir", sep = " ") # or sep = "\t"

txtFile = pd.read_csv("dir")

**Using Open**

txtFile = open("directory, "r)

txtFile2.read()

txtFile2.readline()



In [25]:
textData = pd.read_csv("Documents/Data/flowers.txt")
textData

A: African Daisy
0                B: Bellflower
1               C: Coral Bells
2               D: Desert Rose
3          E: English Bluebell
4             F: Forget Me Not
5                 G: Goldenrod
6                H: Heliotrope
7                 I: Impatiens
8         J: Jamesia americana
9              K: Kangaroo Paw
10       L: Lily of the Valley
11               M: Monks Hood
12                N: Nemophila
13             O: Ox Eye Daisy
14               P: Peace Lily
15            Q: Quaker Ladies
16                R: Rain Lily
17               S: Snapdragon
18             T: Trumpet Vine
19                U: Urn Plant
20       V: Viola wittrockiana
21     W: Whirling Butterflies
22  X: Xanthoceras sorbifolium
23         Y: Yellow Archangel
24           Z: Zinnia elegans

In [26]:
pwd  #shows us the working directory

'C:\\Users\\pinck'

In [27]:
excelFile = pd.read_excel("Downloads/Medicine Junior Doctor template Aug 2020.xlsx", sheet_name = "Rota 2 - B12, Cardio, Haem, MAU")
excelFile.head()

Unnamed: 0           Unnamed: 1 Unnamed: 2       Slot 1    \
0        NaN                  NaN        NaN    Respiratory   
1        NaN                  NaN        NaN     Neeka Muru   
2       Week                 Date        Day            NaN   
3          4  2020-08-05 00:00:00        Wed  08:30 - 17:00   
4        NaN  2020-08-06 00:00:00        Thu  08:15 - 17:00   

                   Slot 2         Slot 3         Slot 4          Slot 5  \
0             Respiratory    Respiratory    Respiratory     Respiratory   
1  Sarfaraz Hussain Shah    Monica Ulotu         VACANT  Victoria Onen    
2                     NaN            NaN            NaN             NaN   
3           08:30 - 17:00  08:30 - 17:00  08:30 - 17:00   08:30 - 17:00   
4           08:15 - 17:00  08:15 - 17:00  08:15 - 17:00   08:15 - 17:00   

                   Slot 6                Slot 7                     Slot 8  \
0             Respiratory            Cardiology                 Cardiology   
1             John Kamara         Stephanie Goh                    Yao Mau   
2                     NaN                   NaN                        NaN   
3  Nights \n21:00 - 09:30  Late \n15:30 - 21:30                        OFF   
4  Nights \n21:00 - 09:30  Late \n15:30 - 21:30   Long Day\n 09:00 - 21:30   

                Slot 9        Slot 10                    Slot 11  \
0           Cardiology     Cardiology                 Cardiology   
1  Efthymia Kallistrou    Dimple Shah               Zainab Aftab   
2                  NaN            NaN                        NaN   
3        09:30 - 17:00  08:30 - 17:00   Long Day\n 09:00 - 21:30   
4        08:15 - 17:00  08:15 - 17:00              08:30 - 17:00   

         Slot 12                  Slot 13                    Slot 14  \
0     Cardiology                 Haem/Onc                   Haem/Onc   
1  Naqibul Islam              Alex Daniel           Renata Risteska    
2            NaN                      NaN                        NaN   
3  08:30 - 17:00  outliers\n08:30 - 17:00   Long Day\n 09:00 - 21:30   
4  08:15 - 17:00  outliers\n08:15 - 17:00              08:15 - 17:00   

                           Slot 15           Slot 16  
0                              AFU               MAU  
1  Bernadett Vago \n(starts 07/09)  Ifeoluwa Gbeleyi  
2                              NaN               NaN  
3                    08:30 - 17:00     08:30 - 17:00  
4                    08:15 - 17:00     08:15 - 17:00

NaN means empty cells not zero

In [28]:
excelFile.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 373 entries, 0 to 372
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  54 non-null     object
 1   Unnamed: 1  365 non-null    object
 2   Unnamed: 2  365 non-null    object
 3   Slot 1      372 non-null    object
 4   Slot 2      372 non-null    object
 5   Slot 3      372 non-null    object
 6   Slot 4      372 non-null    object
 7   Slot 5      372 non-null    object
 8   Slot 6      372 non-null    object
 9   Slot 7      372 non-null    object
 10  Slot 8      372 non-null    object
 11  Slot 9      372 non-null    object
 12  Slot 10     372 non-null    object
 13  Slot 11     371 non-null    object
 14  Slot 12     372 non-null    object
 15  Slot 13     372 non-null    object
 16  Slot 14     372 non-null    object
 17  Slot 15     372 non-null    object
 18  Slot 16     372 non-null    object
dtypes: object(19)
memory usage: 55.5+ KB


In [29]:
text = open("Documents/Data/flowers.txt", "r")
text.read()

'A: African Daisy\nB: Bellflower\nC: Coral Bells\nD: Desert Rose\nE: English Bluebell\nF: Forget Me Not\nG: Goldenrod\nH: Heliotrope\nI: Impatiens\nJ: Jamesia americana\nK: Kangaroo Paw\nL: Lily of the Valley\nM: Monks Hood\nN: Nemophila\nO: Ox Eye Daisy\nP: Peace Lily\nQ: Quaker Ladies\nR: Rain Lily\nS: Snapdragon\nT: Trumpet Vine\nU: Urn Plant\nV: Viola wittrockiana\nW: Whirling Butterflies\nX: Xanthoceras sorbifolium\nY: Yellow Archangel\nZ: Zinnia elegans'

In [30]:
#Data from the web

webF = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data-original")
webF

18.0   8.   307.0      130.0      3504.      12.0   70.  1.\t"chevrolet chevelle malibu"
0    15.0   8.   350.0      165.0      3693.      1...                                      
1    18.0   8.   318.0      150.0      3436.      1...                                      
2    16.0   8.   304.0      150.0      3433.      1...                                      
3    17.0   8.   302.0      140.0      3449.      1...                                      
4    15.0   8.   429.0      198.0      4341.      1...                                      
..                                                 ...                                      
400  27.0   4.   140.0      86.00      2790.      1...                                      
401  44.0   4.   97.00      52.00      2130.      2...                                      
402  32.0   4.   135.0      84.00      2295.      1...                                      
403  28.0   4.   120.0      79.00      2625.      1...                                      
404  31.0   4.   119.0      82.00      2720.      1...                                      

[405 rows x 1 columns]

In [31]:
webF.head()

18.0   8.   307.0      130.0      3504.      12.0   70.  1.\t"chevrolet chevelle malibu"
0  15.0   8.   350.0      165.0      3693.      1...                                      
1  18.0   8.   318.0      150.0      3436.      1...                                      
2  16.0   8.   304.0      150.0      3433.      1...                                      
3  17.0   8.   302.0      140.0      3449.      1...                                      
4  15.0   8.   429.0      198.0      4341.      1...

In [32]:
webF.shape

(405, 1)

In [77]:
# notice the data above has no header, so the first row was printed on top as header
#notice data separated by "\t", so can add it to the argument when opening the csv file

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data-original"
webF2 = pd.read_csv(url, sep = "\t", header = None)  #header= None tells it there is no header, so doesnt move the first row as column
webF2

0  \
0    18.0   8.   307.0      130.0      3504.      1...   
1    15.0   8.   350.0      165.0      3693.      1...   
2    18.0   8.   318.0      150.0      3436.      1...   
3    16.0   8.   304.0      150.0      3433.      1...   
4    17.0   8.   302.0      140.0      3449.      1...   
..                                                 ...   
401  27.0   4.   140.0      86.00      2790.      1...   
402  44.0   4.   97.00      52.00      2130.      2...   
403  32.0   4.   135.0      84.00      2295.      1...   
404  28.0   4.   120.0      79.00      2625.      1...   
405  31.0   4.   119.0      82.00      2720.      1...   

                             1  
0    chevrolet chevelle malibu  
1            buick skylark 320  
2           plymouth satellite  
3                amc rebel sst  
4                  ford torino  
..                         ...  
401            ford mustang gl  
402                  vw pickup  
403              dodge rampage  
404                ford ranger  
405                 chevy s-10  

[406 rows x 2 columns]

In [78]:
webF2.head(15)
# notice column 0 is lots of colums cramped into one,so will need to be split

0  \
0   18.0   8.   307.0      130.0      3504.      1...   
1   15.0   8.   350.0      165.0      3693.      1...   
2   18.0   8.   318.0      150.0      3436.      1...   
3   16.0   8.   304.0      150.0      3433.      1...   
4   17.0   8.   302.0      140.0      3449.      1...   
5   15.0   8.   429.0      198.0      4341.      1...   
6   14.0   8.   454.0      220.0      4354.       ...   
7   14.0   8.   440.0      215.0      4312.       ...   
8   14.0   8.   455.0      225.0      4425.      1...   
9   15.0   8.   390.0      190.0      3850.       ...   
10  NA     4.   133.0      115.0      3090.      1...   
11  NA     8.   350.0      165.0      4142.      1...   
12  NA     8.   351.0      153.0      4034.      1...   
13  NA     8.   383.0      175.0      4166.      1...   
14  NA     8.   360.0      175.0      3850.      1...   

                                   1  
0          chevrolet chevelle malibu  
1                  buick skylark 320  
2                 plymouth satellite  
3                      amc rebel sst  
4                        ford torino  
5                   ford galaxie 500  
6                   chevrolet impala  
7                  plymouth fury iii  
8                   pontiac catalina  
9                 amc ambassador dpl  
10              citroen ds-21 pallas  
11  chevrolet chevelle concours (sw)  
12                  ford torino (sw)  
13           plymouth satellite (sw)  
14                amc rebel sst (sw)

In [79]:
#create copy just incase you run into issues, you dont need to start afresh
webF2copy = webF2.copy()

In [80]:
#splitting Data
# cant use loc since the data is not named

webF2copy.iloc[:, 0]  #retrive column 0

#convert the column to a str
webF2copy.iloc[:, 0].str

#split the string using split method
webF2copy.iloc[:, 0].str.split()

#store the new list in to a new or the same variable
webF2copy = webF2copy.iloc[:, 0].str.split(expand=True)   # expand=true makes it print like a neat data frame, otherwiseprints cramped up

In [81]:
webF2copy.head()

0   1      2      3      4     5    6   7
0  18.0  8.  307.0  130.0  3504.  12.0  70.  1.
1  15.0  8.  350.0  165.0  3693.  11.5  70.  1.
2  18.0  8.  318.0  150.0  3436.  11.0  70.  1.
3  16.0  8.  304.0  150.0  3433.  12.0  70.  1.
4  17.0  8.  302.0  140.0  3449.  10.5  70.  1.

In [82]:
# pd.concat adds 2 tables together normally row-wise. To add them column wise, you need to add axis into the argument 

In [83]:
carName = webF2.iloc[:,1]  #retrieve car name column from main data
carName.head()

0    chevrolet chevelle malibu
1            buick skylark 320
2           plymouth satellite
3                amc rebel sst
4                  ford torino
Name: 1, dtype: object

In [84]:
cleanData = pd.concat([webF2copy, carName], axis = 1)  #y axis = 1. not putting the axis makes the concatenation row-wise by default. Axis 0 is for row.
cleanData.head()

0   1      2      3      4     5    6   7                          1
0  18.0  8.  307.0  130.0  3504.  12.0  70.  1.  chevrolet chevelle malibu
1  15.0  8.  350.0  165.0  3693.  11.5  70.  1.          buick skylark 320
2  18.0  8.  318.0  150.0  3436.  11.0  70.  1.         plymouth satellite
3  16.0  8.  304.0  150.0  3433.  12.0  70.  1.              amc rebel sst
4  17.0  8.  302.0  140.0  3449.  10.5  70.  1.                ford torino

cleanData.columns = headers

headers is a list of strings representing the name of the columns


In [85]:
headers= ["mpg", "cylinders", "displacement", "horsepower", "weight", "acceleration", "model year", "origin", "car name"]
cleanData.columns = headers
cleanData.head()

mpg cylinders displacement horsepower weight acceleration model year  \
0  18.0        8.        307.0      130.0  3504.         12.0        70.   
1  15.0        8.        350.0      165.0  3693.         11.5        70.   
2  18.0        8.        318.0      150.0  3436.         11.0        70.   
3  16.0        8.        304.0      150.0  3433.         12.0        70.   
4  17.0        8.        302.0      140.0  3449.         10.5        70.   

  origin                   car name  
0     1.  chevrolet chevelle malibu  
1     1.          buick skylark 320  
2     1.         plymouth satellite  
3     1.              amc rebel sst  
4     1.                ford torino

In [86]:
cleanData.info()

#note data type here are all object(i.e string) because we converterted the colum to string earlier, meanwhile some are actually floats and ints.
#non-null means its not empty 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   mpg           406 non-null    object
 1   cylinders     406 non-null    object
 2   displacement  406 non-null    object
 3   horsepower    406 non-null    object
 4   weight        406 non-null    object
 5   acceleration  406 non-null    object
 6   model year    406 non-null    object
 7   origin        406 non-null    object
 8   car name      406 non-null    object
dtypes: object(9)
memory usage: 28.7+ KB


1. mpg-continuous
2. cylinders - multi-valued discrete
3. displacement- continous
4. horsepower- continuous
5. weight- continuous
6. acceleration - continuous
7. model year - multi valued discrete
8. origin multi valued discrete
9. car name - unique for each instance

continous = floats, discrete = integers, object = string
This was gotten from the website where the data was gotten

Becuase we changed those columns to str when we were splitting above to clean data, we need to change them back to their normal format listed above

In [87]:
#continous data being changed from object to floats
cleanData[["mpg", "displacement", "horsepower", "weight", "acceleration"]] = cleanData[["mpg", "displacement", "horsepower", "weight", "acceleration"]].astype('float')

ValueError: could not convert string to float: 'NA'

In [89]:
#the above error means some missing values when collecting the datawere filled as "NA". and program cannot convert this
#string to float. So we need to find the column these values and replace them with something valid

cleanData[cleanData.eq("NA").any(1)] #finding the culprits

mpg cylinders displacement horsepower weight acceleration model year  \
10     NA        4.        133.0      115.0  3090.         17.5        70.   
11     NA        8.        350.0      165.0  4142.         11.5        70.   
12     NA        8.        351.0      153.0  4034.         11.0        70.   
13     NA        8.        383.0      175.0  4166.         10.5        70.   
14     NA        8.        360.0      175.0  3850.         11.0        70.   
17     NA        8.        302.0      140.0  3353.          8.0        70.   
38   25.0        4.        98.00         NA  2046.         19.0        71.   
39     NA        4.        97.00      48.00  1978.         20.0        71.   
133  21.0        6.        200.0         NA  2875.         17.0        74.   
337  40.9        4.        85.00         NA  1835.         17.3        80.   
343  23.6        4.        140.0         NA  2905.         14.3        80.   
361  34.5        4.        100.0         NA  2320.         15.8        81.   
367    NA        4.        121.0      110.0  2800.         15.4        81.   
382  23.0        4.        151.0         NA  3035.         20.5        82.   

    origin                          car name  
10      2.              citroen ds-21 pallas  
11      1.  chevrolet chevelle concours (sw)  
12      1.                  ford torino (sw)  
13      1.           plymouth satellite (sw)  
14      1.                amc rebel sst (sw)  
17      1.             ford mustang boss 302  
38      1.                        ford pinto  
39      2.       volkswagen super beetle 117  
133     1.                     ford maverick  
337     2.              renault lecar deluxe  
343     1.                ford mustang cobra  
361     2.                       renault 18i  
367     2.                         saab 900s  
382     1.                    amc concord dl

above the horsepower and mpg columns have NA

so we have to replace it. usually  using the mean or most frequent number in the columns which can be found as follows.

In [90]:
cleanData.describe()

mpg cylinders displacement horsepower weight acceleration model year  \
count    406       406          406        406    406          406        406   
unique   130         5           83         94    356           97         14   
top     13.0        4.        97.00      150.0  1985.         14.5        73.   
freq      20       207           22         22      4           23         40   

       origin    car name  
count     406         406  
unique      3         312  
top        1.  ford pinto  
freq      254           6

most frequent is 22 in horsepower, the highest is 150.

In [110]:
cleanData["horsepower"] = cleanData["horsepower"].replace("NA", "22") 

In [111]:
cleanData[cleanData.eq("NA").any(1)]   #to check the changes. so we have to changes the NA in mpg now left

Empty DataFrame
Columns: [mpg, cylinders, displacement, horsepower, weight, acceleration, model year, origin, car name]
Index: []

In [96]:
cleanData["mpg"] = cleanData["mpg"].replace("NA", "22")

In [97]:
cleanData[cleanData.eq("NA").any(1)] #checking for any NA still not replaced. nothing prints as there no more NA

Empty DataFrame
Columns: [mpg, cylinders, displacement, horsepower, weight, acceleration, model year, origin, car name]
Index: []

In [104]:
cleanData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           406 non-null    float64
 1   cylinders     406 non-null    object 
 2   displacement  406 non-null    float64
 3   horsepower    406 non-null    float64
 4   weight        406 non-null    float64
 5   acceleration  406 non-null    float64
 6   model year    406 non-null    object 
 7   origin        406 non-null    object 
 8   car name      406 non-null    object 
dtypes: float64(5), object(4)
memory usage: 28.7+ KB


In [127]:
#continous data being changed from object to floats. works now as no NA
cleanData[["mpg", "displacement", "horsepower", "weight", "acceleration"]] = cleanData[["mpg", "displacement", "horsepower", "weight", "acceleration"]].astype('float')

In [133]:
#converting the int objects that are currently strings to their rightful int

cleanData[["cylinders", "model year", "origin"]] = cleanData[["cylinders", "model year", "origin"]].astype("float")   #need to turn to float first because there is value "8." in cylinders which will throw error as it wont convertto int directly

In [135]:
cleanData[["cylinders", "model year", "origin"]] = cleanData[["cylinders", "model year", "origin"]].astype("int")

In [137]:
#sorting in alphabetic order
cleanData.sort_values("car name")

mpg  cylinders  displacement  horsepower  weight  acceleration  \
103  13.0          8         360.0       175.0  3821.0          11.0   
9    15.0          8         390.0       190.0  3850.0           8.5   
73   17.0          8         304.0       150.0  3672.0          11.5   
264  19.4          6         232.0        90.0  3210.0          17.2   
322  24.3          4         151.0        90.0  3003.0          20.1   
..    ...        ...           ...         ...     ...           ...   
402  44.0          4          97.0        52.0  2130.0          24.6   
204  29.0          4          90.0        70.0  1937.0          14.2   
316  41.5          4          98.0        76.0  2144.0          14.7   
332  44.3          4          90.0        48.0  2085.0          21.7   
300  31.9          4          89.0        71.0  1925.0          14.0   

     model year  origin                 car name  
103          73       1  amc ambassador brougham  
9            70       1       amc ambassador dpl  
73           72       1       amc ambassador sst  
264          78       1              amc concord  
322          80       1              amc concord  
..          ...     ...                      ...  
402          82       2                vw pickup  
204          76       2                vw rabbit  
316          80       2                vw rabbit  
332          80       2     vw rabbit c (diesel)  
300          79       2         vw rabbit custom  

[406 rows x 9 columns]

In [139]:
cleanData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 406 entries, 0 to 405
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           406 non-null    float64
 1   cylinders     406 non-null    int32  
 2   displacement  406 non-null    float64
 3   horsepower    406 non-null    float64
 4   weight        406 non-null    float64
 5   acceleration  406 non-null    float64
 6   model year    406 non-null    int32  
 7   origin        406 non-null    int32  
 8   car name      406 non-null    object 
dtypes: float64(5), int32(3), object(1)
memory usage: 23.9+ KB


In [140]:
cleanData.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [145]:
cleanData["horsepower"].max()

230.0

In [146]:
cleanData["horsepower"].min()

22.0

In [148]:
cleanData["horsepower"].median()

93.5

In [149]:
cleanData["horsepower"].mean()

103.85467980295566

In [150]:
cleanData["horsepower"].mode()

0    150.0
dtype: float64

In [151]:
cleanData["horsepower"].fillna(round(cleanData["horsepower"].mean(), 1), inplace=True)
# or .fillna(32) if you just want to fill with a simple number

In [ ]:
#dropping any row or column with Nan
data.dropna()
#dropping specific columns with Nan
data.drop(columns=["age"], inplace=True)

In [152]:
cleanData.isnull().sum()

mpg             0
cylinders       0
displacement    0
horsepower      0
weight          0
acceleration    0
model year      0
origin          0
car name        0
dtype: int64

In [158]:
cleanData.head()

mpg  cylinders  displacement  horsepower  weight  acceleration  \
0  18.0          8         307.0       130.0  3504.0          12.0   
1  15.0          8         350.0       165.0  3693.0          11.5   
2  18.0          8         318.0       150.0  3436.0          11.0   
3  16.0          8         304.0       150.0  3433.0          12.0   
4  17.0          8         302.0       140.0  3449.0          10.5   

   model year  origin                   car name  
0          70       1  chevrolet chevelle malibu  
1          70       1          buick skylark 320  
2          70       1         plymouth satellite  
3          70       1              amc rebel sst  
4          70       1                ford torino

EXPORTING THE DATA

In [159]:
cleanData.to_csv("CarData.csv", index=False)#index = false makes sure the data isnts saved with the index because next time you load it, the index will be a new column

In [160]:
cleanData["cylinders"].unique()

array([8, 4, 6, 3, 5])

In [161]:
#Checking for unique categories

#data["column-name"].unique() ---- shows the nmber of unique data in the column

cleanData["cylinders"].value_counts() #----------number of each sample in each category

4    207
8    108
6     84
3      4
5      3
Name: cylinders, dtype: int64